### PR TITLE
Proposal to add some new ordering functions

### PR DESCRIPTION
**Pull request by jmpavlick**
Hey, I've been processing a lot of lists lately; with the help of some of the wonderful people in Elm Slack (particularly @Arkham, who provided me with some prior art that this code is largely based on), I've stumbled across a couple of functions that I feel belong in `elm-community/basics-extra`.

I believe that the code included in this pull request satisfies the requirements listed under the `Contributing` subheading on in the `elm-community/basics-extra` docs, but I felt that it might be helpful to explain myself further in the pull request, as well.

I am proposing to add two functions:

```
orderBy : List (a -> a -> Order) -> (a -> a -> Order)
```

and

```
toOrder : (a -> comparable) -> (a -> a -> Order)
```

`orderBy` allows you to combine multiple `a -> a -> Order` functions into one function. This is useful if you have a type alias, and you'd like to sort a list of values for that type alias, based on multiple properties. 

`toOrder` gives you a shorthand for creating an `a -> a -> Order` function from a function `a -> comparable`. We already have `List.sortBy`, which does this for you for a single-dimension sort; `toOrder` gives you a convenient way to turn multiple property accessors into sortable dimensions, so that you can use them with `orderBy`, then with `List.sortWith`.

Tests and `elm-verify-examples`-powered docs that explain this more thoroughly are included in the pull request.

I feel that these belong in `elm-community/basics-extra`, because `elm/core`'s `Basics` module exposes a `compare` function that returns `Order`, as these functions do. Functions that return `Order` fit with the rest of the functions in `Basics` (as opposed to `List`). I believe that it would be helpful for the community to have access to some more powerful ordering functions, and that this package is the best place for them.

While I was working on this, I version-bumped all of the dependencies in `package.json`, as well as added a few extra dependencies that made my development and documentation-writing process nicer. I added `elm-community/list-extra` as a test dependency so that I could use `List.Extra.permutations` to make sure that all possible permutations of a given `List a` sorted as expected, across multiple dimensions. I understand if the maintainers object to any of these additions, and I am not opposed to removing them if requested to do so (or to in-lining the definition of `List.Extra.permutations` into the `Tests.elm` module, to remove that dependency).

If there is anything else that I need to do in order to make it so that this PR is mergeable, or if you have any questions or comments, please let me know!

Thanks,
John

### CODE REVIEW

**Comment by jmpavlick** [comment]
For completeness' sake, I've also added a function `toOrderDesc`, that performs the same operation as `toOrder`, with flipped comparisons (i.e., `LT` becomes `GT`, and vice-versa).

File: elm.json
Comment (Janiczek) [inline elm.json]: I'm aware this PR is already doing stuff far above just adding two functions, but I just wanted to note that this could be "2.0.1 <= v < 3.0.0" now, if we upgrade the `elm-test` npm package as well :)

File: src/Basics/Extra.elm
Comment (Janiczek) [inline src/Basics/Extra.elm]: There is a @docs entry missing for `toOrderDesc`

File: src/Basics/Extra.elm
Comment (Janiczek) [inline src/Basics/Extra.elm]: Something like this would be a short-circuiting version, ie. not traversing all the comparators if we've already got the LT or GT result:
```suggestion
       case comparators of
           [] ->
               EQ
               
           comparator :: rest ->
               case comparator a b of
                   EQ ->
                       orderBy rest a b
                       
                   other ->
                       other
```

File: elm.json
Comment (jmpavlick) [reply elm.json]: Fixed, thank you!

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: Wow, I really like this - much easier to follow.

I felt like there was a simpler solution than using a `foldl`, but I wasn't quite there. Thank you!

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: Good catch, thank you!

File: package.json
Comment (Janiczek) [inline package.json]: We've upgraded `elm-explorations/test` in `elm.json` but this needs us to also upgrade `elm-test` NPM dependency:
Screenshot 2022-11-16 at 16 59 54
```suggestion
    "elm-test": "^0.19.1-revision10"
```

Comment (jmpavlick) [reply package.json]: Oh whoops, I ran `npm upgrade` and apparently it did not do what I expected it to do. Thank you!

**Comment by jmpavlick** [comment]
Okay @Janiczek - _now_ `package-lock.json` is fixed, too 😄

**Comment by jmpavlick** [comment]
TODO: Add `elm-coverage` npm package, and make sure that all functions are fully covered.

**Comment by jmpavlick** [comment]
Everything has full test coverage.

I know that there are already doctests for a lot of stuff, but I felt like it was worth doing anyway.

**Comment by jmpavlick** [comment]
Made a small fix to make it so that `orderBy` could be tail-call optimized.

File: src/Basics/Extra.elm
Comment (pzp1997) [inline src/Basics/Extra.elm]:
```suggestion
toOrderDesc selector a b =
    Basics.compare (selector b) (selector a)
```

File: src/Basics/Extra.elm
Comment (pzp1997) [inline src/Basics/Extra.elm]:
```suggestion
toOrder selector a b =
    Basics.compare (selector a) (selector b)
```

File: src/Basics/Extra.elm
Comment (pzp1997) [inline src/Basics/Extra.elm]: Should this be `colorToComparable`?

```suggestion
            , toOrder (.color >> colorToComparable)
```

**Review by pzp1997 — APPROVED** [review]
Approve with minor changes

**Comment by jmpavlick** [comment]
Hey @pzp1997 - just saw this. I'll get the recommended changes patched in shortly. Thank you so much!!

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: It absolutely should be - thank you for catching this.

I didn't notice it because while the doctests all claim to pass, for some reason this doctest is being missed. I'm going to commit your suggested change as-is, but I'm also going to take some time to try to figure out why the doctest coverage isn't working as expected.

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: Much cleaner implementation. Thank you for this.

**Comment by jmpavlick** [comment]
@pzp1997 - I committed all of your proposed changes. I'm not sure why the doc tests didn't catch the `colorToOrder` / `colorToComparable` bug, but I verified that your proposed change works (and that mine doesn't!).

Really appreciate you taking the time to stop back by this PR to leave feedback. I can't merge because I don't have write access, so you can feel free to merge this in at your leisure, if you'd like.

Thanks again!

File: src/Basics/Extra.elm
Comment (pzp1997) [inline src/Basics/Extra.elm]: Delete lines 406-407

File: src/Basics/Extra.elm
Comment (pzp1997) [inline src/Basics/Extra.elm]: Delete lines 437-446

**Comment by jmpavlick** [comment]
@pzp1997 sorry, missed those changes - will fix in a moment.

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: Fixed!

File: src/Basics/Extra.elm
Comment (jmpavlick) [reply src/Basics/Extra.elm]: Fixed!

**Comment by jmpavlick** [comment]
@pzp1997 done, sorry about that - thank you!

**Review by pzp1997 — APPROVED** [review]

**Pull request merged**
